### ISSUE TITLE
[HF-4.0] Autolykos 2.0 implementation - memory hardness

### ISSUE THREAD

**Issue by kushti** [ergoplatform/ergo#1100]
Based on proposal by BigEvilSoloMiner

Changes to the PoW whitepaper:

Algorithm 1 - Block mining: 
Line 3: Change from  H(j|M|pk|m|w) to H(j|pk|w|M|m)
Line 6: Change from genIndexes(m|nonce) to genIndexes (j|pk|w|M|m|nonce)
Line 6: j is computed as (m * nonce) modulo N

Algorithm 2 - Solution verification: 
Line 4: Change from genIndexes(m|nonce) to genIndexes(j|pk|w|M|m|nonce)
Line 4: j is computed as (m * nonce) modulo N

They should make Autolykos memory-harder. Please note non-outsoutsourceability not touched by the proposed changes

**Comment by jasondavies** [comment]
A note on credits: although BigEvilSoloMiner found a way to avoid memory-hardness by iterating over `w` for a constant `nonce`, their original fix was incorrect and I actually suggested the correct fix, which is to use the hash:

```
H(j|pk|w|M|m)
```

This prevents the miner from iterating over `w` as it is placed *before* the large data blob `M` making it expensive to change `w`.

I don't see any point in modifying `genIndexes` without a specific motivation.  (Too many changes at once make it harder to analyse, and I'm suspicious of making any changes without specifying why they are made.)

**Comment by AsicResearcher** [comment]
**genIndexes** must be memory bound as well because we want the compute cost for bypassing it to be extreme high, also it closes the door on any potential mathematical shortcuts.

**Comment by jasondavies** [comment]
> genIndexes must be memory bound as well because we want the compute cost for bypassing it to be extreme high, also it closes the door on any potential mathematical shortcuts.

The proposed change is to replace `genIndexes(m|nonce)` with `genIndexes(j|pk|w|M|m|nonce)`, with the additional property: `j = (m * nonce) mod N`.

However, this has the (perhaps unintentional) side-effect of making mining even worse for lower-memory GPUs, since they now need to cache double the number of hashes compared with Autolykos V1: 2^26 hashes for `H(j|pk|w|M|m)` and another 2^26 hashes for `H(j|pk|w|M|m|nonce)` (for the proposed genIndexes change).  This means a total of 4GB instead of 2GB needs to be allocated for the hashes alone.

Your argument is that you want the cost of "bypassing it to be extremely high", but I don't see any reason for genIndexes to be memory-bound in the first place, therefore there isn't anything to bypass here.  The only property it should have is generating uniformly random indexes, and I think your proposed change might actually make it possible to subvert this property (see below).

> also it closes the door on any potential mathematical shortcuts.

Not necessarily.  Adding more complexity could potentially mean *more* room for shortcuts.  That's why the proof-of-work should ideally be simple enough to analyse properly.

In fact, I think the proposed change *does* have a mathematical shortcut that allows the uniformly-random property of the proposed `genIndexes` to be subverted: since the miner controls `nonce`, they can pick nonces with particular properties that influence `j` so that memory accesses are not uniformly random.

In summary, I really don't think there is any reason for `genIndexes` to be memory-bound in the first place, and these proposed changes seem like they make it easier to subvert what it was designed to do in the first place.

**Comment by kushti** [comment]
@jasondavies @AsicResearcher concluding, the least common denominator is about just replacing _H(j|M|pk|m|w)_ with _H(j|pk|w|M|m)_

**Comment by jasondavies** [comment]
This should restore memory-hardness, yes.

**Comment by kushti** [comment]
Proposed changes are finalized in https://github.com/ergoplatform/eips/pull/22 , implementation is in #1141 .
